solr/core/src/java/org/apache/solr/core/HttpSolrClientProvider.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.solr.core;
+
+import static org.apache.solr.util.stats.InstrumentedHttpRequestExecutor.KNOWN_METRIC_NAME_STRATEGIES;
+
+import java.lang.invoke.MethodHandles;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import org.apache.solr.client.solrj.SolrClient;
+import org.apache.solr.client.solrj.impl.Http2SolrClient;
+import org.apache.solr.client.solrj.impl.HttpClientUtil;
+import org.apache.solr.common.SolrException;
+import org.apache.solr.common.util.IOUtils;
+import org.apache.solr.metrics.SolrMetricManager;
+import org.apache.solr.metrics.SolrMetricsContext;
+import org.apache.solr.security.HttpClientBuilderPlugin;
+import org.apache.solr.update.UpdateShardHandlerConfig;
+import org.apache.solr.util.stats.HttpClientMetricNameStrategy;
+import org.apache.solr.util.stats.InstrumentedHttpListenerFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Internal use only. Provider of the default SolrClient implementation. */
```

Comment:
flip the sentence order.  Javadocs should always start with the summary of what it is/does.  Internal Use only comment can be replaced with `@lucene.internal` which we use a lot and probably should use more.  But it's not that pertinent outside SolrJ.   Glad you were thinking of documenting internal-ness.  Maybe the class doesn't need to be public, which would imply its internal-ness without even needing an annotation/docs.

Reply:
Yeah agreed to make it package private. I guess I wanted to use it in UpdateShardHandler as well, like moving clients to this class. But yeah let's not make it too much complex for now. 

But in future, I wonder sometimes, Do we even need multiple http clients -- default, update or recovery?In past,  due to some issues with http1 we created multiple, however, is that reason even valid anymore with http2? If not, there is a possibility that having one client could improve the performance.  I guess benchmarking could have the answer to it.

Reply:
The different ones were added intentionally to tune timeouts, which matters at high scale when Solr is pushed to extremes.

Reply:
We can always re-set connection/socket timeout by re-creating the Http2SolrClient but using the same jetty client. Or is there any other setting you are talking about? I mean yes we do have many different settings at Jetty level as well, however looks like so far everything is similar for all three Http2SolrClients. 

I remember coming across a comment from Mark when the new code update client was introduced. 

https://issues.apache.org/jira/browse/SOLR-12293

> Currently the pool is shared too broadly - for example during replication we don't guarantee we read the full streams when downloading index files and we don't necessarily want to, emptying the stream for a huge file due to error or abort is too expensive. We can't have these connections pollute the update connection pool.

Reply:
That's a great reason to keep it separate, even if the configuration is identical.